### ISSUE TITLE
How can I determine if an input image is defected or not?

### ISSUE THREAD

**Issue by lamnguyenvux** [JohnnyHopp/PaDiM-EfficientNet_NS-and-PaDiM-EfficientNetV2#2]
After calculating mean and cov for each patch(x, y) in training set, I feed an image through pretrained model, take out feature maps of each layer, concatenate them then calculate distance to each patch distribution. Basically, I will have the `score_map`, then up-sample it and normalize it. After that, how can I tell if input data is defected or not? Assume that I already have a optimal threshold. I compare this score_map to the threshold. If any pixel of  `score_map` is bigger than the threshold value, input data will be predicted as defect. In the other hand, if all pixels of `score_map` are smaller than threshold value, it will be predicted as normal. Am I correct? If not, please let me know. Many thanks!

**Comment by JohnnyHopp** [comment]
@lamnguyenvu98 Hi, your understanding is totally correct. ^_^

**Comment by lamnguyenvux** [comment]
Hello again,

I appreciate your response. Currently, I'm trying to make an inference script to detect defect in from a single image or from camera. I tested with few samples from dataset. Defected samples worked quiet well but for normal samples, the model always predict a mask, which had many pixels of value `1`. It was supposed to be a black mask. I even used normal image from training set but result was the same.

Here is one image which is taken from training set. As you can see, it has many pixels of value `1`. Based on the logic I mentioned above, this image will be predicted as defect.

![image](https://user-images.githubusercontent.com/83378878/216776667-309a3502-4ab0-4e48-90a7-177b17d65d86.png)
 
Here is my `predict` function:

```
    def predict(self, data):
        '''
            data [torch.Tensor]: input data which has shape (batch, C, H, W)
                H, W    = (224, 224)
                C       = 3
                batch   = 1
        '''
        self.backbone.eval()
        
        outputs = OrderedDict([('layer1', []), ('layer2', []), ('layer3', [])])
        
        tfms = transforms.Normalize(
                    mean=[0.485, 0.456, 0.406],
                    std=[0.229, 0.224, 0.225]
                )
        
        data = tfms(data)
        
        with torch.no_grad():
            _ = self.backbone(data.to(self.device))
        
        for k, v in zip(outputs.keys(), self.output_layers):
            outputs[k].append(v.cpu().detach())

        self.output_layers.clear()
        
        for k, v in outputs.items():
            outputs[k] = torch.cat(v, 0)
        
        # Embedding concat
        embedding_vectors = outputs['layer1']
        for layer_name in ['layer2', 'layer3']:
            embedding_vectors = embedding_concat(embedding_vectors, outputs[layer_name])
        
        # randomly select d dimension
        embedding_vectors = torch.index_select(embedding_vectors, 1, self.idx)
        # calculate distance matrix
        B, C, H, W = embedding_vectors.size()
        embedding_vectors = embedding_vectors.view(B, C, H * W).numpy()
        dist_list = []
        for i in range(H * W):
            mean = self.distribution[0][:, i]
            conv_inv = np.linalg.inv(self.distribution[1][:, :, i])
            dist = [mahalanobis(sample[:, i], mean, conv_inv) for sample in embedding_vectors]
            dist_list.append(dist)

        dist_list = np.array(dist_list).transpose(1, 0).reshape(B, H, W)

        # upsample
        dist_list = torch.tensor(dist_list)
        score_map = F.interpolate(dist_list.unsqueeze(1), size=self.input_size[0], mode='bilinear',
                                    align_corners=False).squeeze(1).numpy()

        # apply gaussian smoothing on the score map
        for i in range(score_map.shape[0]):
            score_map[i] = gaussian_filter(score_map[i], sigma=4)

        # Normalization
        max_score = score_map.max()
        min_score = score_map.min()
        scores = (score_map - min_score) / (max_score - min_score)

        return scores
```
The `self.distribution` is just `[mean, cov]` calculated from training data. Later on, `scores` will be compared to a threshold value to get the mask.

I think my training has some problems but the model predicts mask for defected images quiet well. Did your model work fine with normal images? I just wonder.

**Comment by JohnnyHopp** [comment]
> Hello again,
> 
> I appreciate your response. Currently, I'm trying to make an inference script to detect defect in from a single image or from camera. I tested with few samples from dataset. Defected samples worked quiet well but for normal samples, the model always predict a mask, which had many pixels of value `1`. It was supposed to be a black mask. I even used normal image from training set but result was the same.
> 
> Here is one image which is taken from training set. As you can see, it has many pixels of value `1`. Based on the logic I mentioned above, this image will be predicted as defect.
> 
> ![image](https://user-images.githubusercontent.com/83378878/216776667-309a3502-4ab0-4e48-90a7-177b17d65d86.png)
> 
> Here is my `predict` function:
> 
> ```
>     def predict(self, data):
>         '''
>             data [torch.Tensor]: input data which has shape (batch, C, H, W)
>                 H, W    = (224, 224)
>                 C       = 3
>                 batch   = 1
>         '''
>         self.backbone.eval()
>         
>         outputs = OrderedDict([('layer1', []), ('layer2', []), ('layer3', [])])
>         
>         tfms = transforms.Normalize(
>                     mean=[0.485, 0.456, 0.406],
>                     std=[0.229, 0.224, 0.225]
>                 )
>         
>         data = tfms(data)
>         
>         with torch.no_grad():
>             _ = self.backbone(data.to(self.device))
>         
>         for k, v in zip(outputs.keys(), self.output_layers):
>             outputs[k].append(v.cpu().detach())
> 
>         self.output_layers.clear()
>         
>         for k, v in outputs.items():
>             outputs[k] = torch.cat(v, 0)
>         
>         # Embedding concat
>         embedding_vectors = outputs['layer1']
>         for layer_name in ['layer2', 'layer3']:
>             embedding_vectors = embedding_concat(embedding_vectors, outputs[layer_name])
>         
>         # randomly select d dimension
>         embedding_vectors = torch.index_select(embedding_vectors, 1, self.idx)
>         # calculate distance matrix
>         B, C, H, W = embedding_vectors.size()
>         embedding_vectors = embedding_vectors.view(B, C, H * W).numpy()
>         dist_list = []
>         for i in range(H * W):
>             mean = self.distribution[0][:, i]
>             conv_inv = np.linalg.inv(self.distribution[1][:, :, i])
>             dist = [mahalanobis(sample[:, i], mean, conv_inv) for sample in embedding_vectors]
>             dist_list.append(dist)
> 
>         dist_list = np.array(dist_list).transpose(1, 0).reshape(B, H, W)
> 
>         # upsample
>         dist_list = torch.tensor(dist_list)
>         score_map = F.interpolate(dist_list.unsqueeze(1), size=self.input_size[0], mode='bilinear',
>                                     align_corners=False).squeeze(1).numpy()
> 
>         # apply gaussian smoothing on the score map
>         for i in range(score_map.shape[0]):
>             score_map[i] = gaussian_filter(score_map[i], sigma=4)
> 
>         # Normalization
>         max_score = score_map.max()
>         min_score = score_map.min()
>         scores = (score_map - min_score) / (max_score - min_score)
> 
>         return scores
> ```
> 
> The `self.distribution` is just `[mean, cov]` calculated from training data. Later on, `scores` will be compared to a threshold value to get the mask.
> 
> I think my training has some problems but the model predicts mask for defected images quiet well. Did your model work fine with normal images? I just wonder.

Hi, @lamnguyenvu98, after looking into your code and in my experience, I think the issue might lie in the normalization.
```
        # Normalization
        max_score = score_map.max()
        min_score = score_map.min()
        scores = (score_map - min_score) / (max_score - min_score)
```
Here you should use the max_score and min_score calculated from the whole test set instead of single test image.

I hope it solve your problem and I'd really appreciate it if you can give my repo a star. ^_^

**Comment by lamnguyenvux** [comment]
Thanks again for your response. So it won't work if I only use an image to feed into this algorithm. I guess a batch of images might give better result. Am I right?

**Comment by JohnnyHopp** [comment]
It works. In your case, it's actually a real-world test inference. So at first you need to evaluate on the whole test dataset, then obtain and save max_socre, min_score and optimal threshold for practical inference(any image from real world, test or train dataset should work, single or batch image makes no difference). Remember these parameters are based on the whole test datset instead of each inference image.

**Comment by lamnguyenvux** [comment]
> It works. In your case, it's actually a real-world test inference. So at first you need to evaluate on the whole test dataset, then obtain and save max_socre, min_score and optimal threshold for practical inference(any image from real world, test or train dataset should work, single or batch image makes no difference). Remember these parameters are based on the whole test datset instead of each inference image.

I understand. Thank you so much
